src/distributions/integer.rs

```diff
@@ -109,53 +107,80 @@ impl_nzint!(NonZeroU64, NonZeroU64::new);
 impl_nzint!(NonZeroU128, NonZeroU128::new);
 impl_nzint!(NonZeroUsize, NonZeroUsize::new);
 
-#[cfg(feature = "simd_support")]
-macro_rules! simd_impl {
-    ($(($intrinsic:ident, $vec:ty),)+) => {$(
-        impl Distribution<$intrinsic> for Standard {
-            #[inline]
-            fn sample<R: Rng + ?Sized>(&self, rng: &mut R) -> $intrinsic {
-                $intrinsic::from_bits(rng.gen::<$vec>())
-            }
-        }
-    )+};
 
-    ($bits:expr,) => {};
-    ($bits:expr, $ty:ty, $($ty_more:ty,)*) => {
-        simd_impl!($bits, $($ty_more,)*);
+// Useful for implementations for SIMD types on stable Rust where we cannot use
+// packed_simd's to_le.
+pub(crate) trait SampleNativeEndian {
+    /// Generate a native endian random value of `T`, using `rng` as the source of randomness.
+    fn sample_ne<R: Rng + ?Sized>(rng: &mut R) -> Self;
+}
 
-        impl Distribution<$ty> for Standard {
-            #[inline]
-            fn sample<R: Rng + ?Sized>(&self, rng: &mut R) -> $ty {
-                let mut vec: $ty = Default::default();
-                unsafe {
-                    let ptr = &mut vec;
-                    let b_ptr = &mut *(ptr as *mut $ty as *mut [u8; $bits/8]);
-                    rng.fill_bytes(b_ptr);
+macro_rules! ne_impl {
+    ($($ty:ty),+) => {
+        $(
+            impl SampleNativeEndian for $ty {
+                #[inline]
+                fn sample_ne<R: Rng + ?Sized>(rng: &mut R) -> Self {
+                    let mut vec: MaybeUninit<Self> = MaybeUninit::uninit();
+                    unsafe {
+                        let raw_ptr = vec.as_mut_ptr();
+                        let b_ptr = &mut *(raw_ptr as *mut [u8; mem::size_of::<$ty>()]);
+                        rng.fill_bytes(b_ptr);
+                        vec.assume_init()
+                    }
```

Comment:
This is unsafe, and not just because of the PTR cast: you're passing a `&mut` to uninitialised data into a **safe** method. This is not memory safe, and since we don't have `fill_bytes_uninit` I don't think we have any alternative but to initialise those bytes first.

There was some discussion about such a method in the past (#1080), but as I said then: I expect the performance penalty of initialising first is small.

Reply:
@joshlf @newpavlov can wew do this without `unsafe`?

Reply:
Yeah, it's quite annoying and one of the reasons why I think it would've been nice if Rust had `&out` references.

>can wew do this without unsafe?

I don't think so. We need `unsafe` at the very least for casting array to a SIMD type.

Reply:
This is unsound unfortunately. `b_ptr` is a reference which points to uninitialized memory, which is instantly undefined behavior. Even if that weren't the case, you would also need a guarantee that the memory is only ever written, never read, and `Rng::fill_bytes` gives you no such guarantee.

I would need to think about this more to guarantee that it's sound, but in `zerocopy`, we might be able emit impls of `FromBytes` and `AsBytes` for some SIMD types. That, combined with the recently-added and soon-to-be-released-on-crates.io [`transmute`](https://fuchsia-docs.firebaseapp.com/rust/zerocopy/macro.transmute.html) macro, should allow you to write something like this:

```rust
let mut vec = [0u8; mem::size_of::<$ty>()];
rng.fill_bytes(&mut vec[..]);
transmute!(vec)
```

Note that this initializes `vec` (although it may get optimized out in some cases), but given the API of `Rng::fill_bytes` (both that `Rng` is a safe trait and that `fill_bytes` takes a reference rather than a pointer), you'll never be able to get away from initializing first.

Reply:
We could have another trait where
* `fill_bytes` works with uninitialized data,
* `fill_bytes` is unsafe to implement.

However, I'm not sure this is worth the effort.

Reply:
That will also introduce difficult-to-get-right APIs that could be a big source of unsoundness. If you're worried about the initialization cost, maybe run some of this through Godbolt? My guess is that with simple RNG implementations, the optimizer is probably smart enough to elide the initialization (and with complex RNG implementations, the cost of initialization is going to be negligible anyway).

Reply:
Adding SIMD support to zerocopy [here](fxrev.dev/573083). That look like it'll work for your use case?

Reply:
Also see #1080.

Reply:
Ah. My bad. I was mostly trying to avoid cluttering the code with initializations for different infrastructures.

Reply:
Probably no longer relevant, but I just released SIMD support and the `transmute` macro in zerocopy 0.6.0.